### PR TITLE
Handle Exceptions in building `CallbackContext`

### PR DESCRIPTION
**Pull request by Bibo-Joshi**
Explicitly catches and logs exceptions in creating Callback context objects. Relevant particularly if the user uses a custom class that overrides from_update. Since from_error calls from_update, exceptions in from_update can currently completely shut down update processing *and* error handling. See https://t.me/pythontelegrambottalk/253238 for reference.
Work in progress, e.g. tests are still missing.

### CODE REVIEW

File: telegram/ext/_application.py [hunk 1]
```diff
@@ -1251,30 +1251,43 @@ async def process_update(self, update: object) -> None:
             try:
                 for handler in handlers:
                     check = handler.check_update(update)  # Should the handler handle this update?
-                    if not (check is None or check is False):  # if yes,
-                        if not context:  # build a context if not already built
+                    if check is None or check is False:
```
Comment (harshil21) [inline telegram/ext/_application.py:1254]:
```suggestion
                    if not check:
```

File: telegram/ext/_jobqueue.py [hunk 2]
```diff
@@ -953,7 +955,16 @@ async def _run(
         self, application: "Application[Any, CCT, Any, Any, Any, JobQueue[CCT]]"
     ) -> None:
         try:
-            context = application.context_types.context.from_job(self, application)
+            try:
+                context = application.context_types.context.from_job(self, application)
+            except Exception as exc:
+                _LOGGER.critical(
```
Comment (harshil21) [inline telegram/ext/_jobqueue.py:961]: codecov shows this line wasn't covered

File: tests/ext/test_application.py [hunk 3]
```diff
@@ -2421,3 +2421,81 @@ async def callback(update, context):
...
+            record = caplog.records[0]
+            assert record.name == "telegram.ext.Application"
+            assert record.getMessage().startswith(
+                "Error while building CallbackContext for update 1"
+            )
+            assert record.levelno == logging.CRITICAL
+
+            caplog.clear()
+            with caplog.at_level(logging.CRITICAL):
+                await app.process_update(2)
```
Comment (harshil21) [inline tests/ext/test_application.py:2459]:
```suggestion
            await app.process_update(2)
```
shouldn't the critical warning only be raised when `1` was passed?

File: tests/ext/test_application.py [hunk 4]
```diff
@@ -2421,3 +2421,81 @@ async def callback(update, context):
...
+            record = caplog.records[0]
+            assert record.name == "telegram.ext.Application"
+            assert record.getMessage().startswith(
+                "Error while building CallbackContext for exception 1"
+            )
+            assert record.levelno == logging.CRITICAL
+
+            caplog.clear()
+            with caplog.at_level(logging.CRITICAL):
+                await app.process_error(update=None, error=2)
```
Comment (harshil21) [inline tests/ext/test_application.py:2498]:
```suggestion
            await app.process_error(update=None, error=2)
```
same

File: tests/ext/test_jobqueue.py [hunk 5]
```diff
@@ -646,3 +646,44 @@ async def test_from_aps_job_missing_reference(self, job_queue):
...
+
+    async def test_run_job_exception_in_building_context(
+        self, monkeypatch, job_queue, caplog, app
+    ):
+        # Makes sure that exceptions in building the context don't stop the application
+        exception = ValueError("TestException")
+        original_from_job = CallbackContext.from_job
+
+        def raise_exception(job, application):
+            print(job, job.data, application)
```
Comment (harshil21) [inline tests/ext/test_jobqueue.py:658]:
```suggestion
```

File: tests/ext/test_jobqueue.py [hunk 6]
```diff
@@ -646,3 +646,44 @@ async def test_from_aps_job_missing_reference(self, job_queue):
...
+        assert record.name == "telegram.ext.JobQueue"
+        assert record.getMessage().startswith(
+            "Error while building CallbackContext for job job_callback"
+        )
+        assert record.levelno == logging.CRITICAL
+
+        caplog.clear()
+        with caplog.at_level(logging.CRITICAL):
+            job_queue.run_once(job_callback, 0.1, data=2)
+            await asyncio.sleep(0.2)
```
Comment (harshil21) [inline tests/ext/test_jobqueue.py:686]:
```suggestion
        job_queue.run_once(job_callback, 0.1, data=2)
        await asyncio.sleep(0.2)
```

**Review by harshil21 — COMMENTED** [review]
interesting edge case scenario

File: telegram/ext/_application.py (hunk 1 above)
Comment (Bibo-Joshi) [reply telegram/ext/_application.py:1254]: No, we can't change this. We explicitly document that `None` or `False` indicate that the handler should not trigger but anything else will be passed to `handle_update` https://docs.python-telegram-bot.org/en/v21.1.1/telegram.ext.basehandler.html#telegram.ext.BaseHandler.check_update. Users could be returning `[]` or some other custom object that evaulates to falsy.

File: telegram/ext/_jobqueue.py (hunk 2 above)
Comment (Bibo-Joshi) [reply telegram/ext/_jobqueue.py:961]: That's weird. I blame caching. you even left a comment on the test that covers this :D 

image

File: tests/ext/test_application.py (hunk 3 above)
Comment (Bibo-Joshi) [reply tests/ext/test_application.py:2459]: yes, that's the point here. I'm checking that no log is issued when there is no exception :) I'll leave a comment

File: tests/ext/test_application.py (hunk 4 above)
Comment (Bibo-Joshi) [reply tests/ext/test_application.py:2498]: same as above

File: tests/ext/test_jobqueue.py (hunk 6 above)
Comment (Bibo-Joshi) [reply tests/ext/test_jobqueue.py:686]: Same as above

File: telegram/ext/_application.py (hunk 1 above)
Comment (harshil21) [reply telegram/ext/_application.py:1254]: hmm I suspected that but didn't bother to check the docs, my bad.

File: tests/ext/test_application.py (hunk 3 above)
Comment (harshil21) [reply tests/ext/test_application.py:2460]: ah that's a bit counterintuitive

**Review by harshil21 — APPROVED** [review]
